Add c_flags and cxx_flags to env profile settings

gretay-js opened the pull request:
Add support for c_flags and cxx_flags fields in profile settings of env stanza.  This is in addition to existing ocaml flags in env. It enables us to specify compilation flags per profile when building libraries with C or C++ stubs.

rgrinberg commented on src/env_node.ml:
```diff
@@ -109,3 +113,53 @@ let rec ocaml_flags t ~profile ~expander =
+    t.c_flags <- Some flags;
+    flags
+
+let rec cxx_flags t ~profile ~expander =
```
This function looks awfully similar to `c_flags`. Would you mind trying to generalize them to with some sort of `make_flags` function? It would probably require passing some functions as parameters but it's worth it in this case I think.

gretay-js replied:
This kind of recursive search through directories is a recurrent pattern in env_node and it is used by other existing functions (ocaml_flags, local_binaries, artifcats).  I am not sure how to best fix it. Wouldn't it be better to fix this in a separate PR?

rgrinberg commented on src/super_context.mli:
```diff
@@ -53,6 +53,20 @@ val ocaml_flags
   -> Buildable.t
   -> Ocaml_flags.t
 
+val c_flags
```
In fact, it would be better if these flags functions just took expanders. We're trying to minimize uses of Super_context as it's a large, and mostly meaningless module at this point.

gretay-js replied:
fixed.

rgrinberg commented on src/lib_rules.ml:
```diff
@@ -287,15 +287,15 @@ module Gen (P : Install_rules.Params) = struct
+      build_c_file   lib ~dir ~includes (resolve_name name ~ext:".c")
     ) @ List.map lib.cxx_names ~f:(fun name ->
-      build_cxx_file lib ~expander ~dir ~includes (resolve_name name ~ext:".cpp")
+      build_cxx_file lib ~dir ~includes (resolve_name name ~ext:".cpp")
```
Passing expanders is preferable. It's a bit tedious at the moment but we're trying to minimize uses of Super_context.

gretay-js replied:
fixed.

rgrinberg commented on test/blackbox-tests/test-cases/env-cflags/dune-project:
```diff
@@ -0,0 +1 @@
+(lang dune 1.0)
```
This will need to be 1.7 as you add the appropriate Syntax.since

gretay-js replied:
indeed, it does, which means I added Syntax.since correctly. fixed.

rgrinberg commented on src/super_context.ml:
```diff
@@ -237,10 +247,52 @@ let ocaml_flags t ~dir (x : Buildable.t) =
+  end
+
+
+let cxx_flags_gather t ~dir ~(lib : Library.t) ccg =
```
Why do we need the `_gather` suffix here?

gretay-js replied:
There is another cxx_flags function in Super_context, and a corresponding field of t.  The existing function inspects context.ocamlc_flags and from them computes cxx_flags. Would it be better to rename existing cxx_flags to something like cxx_flags_from_ocamlc_flags or cxx_flags_default?

rgrinberg replied:
I think that it would be better to remove the original function. It seems like we shouldn't really be using it anymore as the `cxx_flags` are now per directory.

gretay-js replied:
The original cxx_flags has been used in combination with expanded lib.cxx_flags in lib_rules.ml to build_cxx_file. The new cxx_flags_gather replaces lib.cxx_flags. I am not sure why it makes the original cxx_flags obsolete. For that matter, I don't know why the original cxx_flags were computed in this way from ocamlc_flags and what guarantees that C++ compiler knows about all possible ocamlc_flags (except -std) used in a project. The must be some assumptions about the use of these cxx_flags that I am missing.

In any case, I attempted to make this change in two steps:
1) change Super_context to rename the original cxx_flags to cxx_flags_orig and renamed cxx_flags_gather to cxx_flags.
55a4d224be5210ffeba7c551b4ad7bb2bd1a745b
2) remove cxx_flags_orig from Super_context.  
They are no longer used anywhere except in Pform.create, where they are getting an akward special treatment. I replaced it by recomputing them (as originally in super_context) and it is analogous to the way cflags are obtained in Pform.
61045a521d6646b5a44780e2fab3c39cd71f8c75

My main question is whether cxx_flags_orig need to be conjoined to the new cxx_flags or simply dropped from build_cxx_files in lib_rule. For now, I dropped them, but it might not be safe: removing the original cxx_flags might break other things in dune or external projects that rely on these flags being inferred from ocamlc_flags.

There is a test in the testsuite for it that passes. I haven't had a chance to test it in a project that actually compiles, because the stubs in my project are in C, but I can create such a test, if this solution seems good to you.

gretay-js replied:
Thank you for the explanations! It makes sense. I try to add a test for (4).

gretay-js replied:
> > My main question is whether cxx_flags_orig need to be conjoined to the new cxx_flags or simply dropped from build_cxx_files in lib_rule. For now, I dropped them, but it might not be safe: removing the original cxx_flags might break other things in dune or external projects that rely on these flags being inferred from ocamlc_flags.
> 
> Those are the defaults so they should be inherited. If a user wants to drop them, then setting them explicitly is always possible (without the use of `:standard`).

It looks like  cxx_flags_orig was always added in lib_rules, regardless of ":standard". 
After removing this default in commit 61045a521d6646b5a44780e2fab3c39cd71f8c75, I have noticed that one package from opam fails to build with this version.
The package is re2 and the problem is in 
https://github.com/janestreet/re2/blob/master/src/jbuild.
Adding -fPIC at least on my machine fixes the problem, not sure if that's a good solution in general.

gretay-js replied:
> It looks like cxx_flags_orig was always added in lib_rules, regardless of ":standard".
> After removing this default in commit [61045a5](https://github.com/ocaml/dune/commit/61045a521d6646b5a44780e2fab3c39cd71f8c75), I have noticed that one package from opam fails to build with this version.
> The package is re2 and the problem is in
> https://github.com/janestreet/re2/blob/master/src/jbuild.
> Adding -fPIC at least on my machine fixes the problem, not sure if that's a good solution in general.

This is now fixed. Thanks to @xclerc for help!

rgrinberg commented on src/env_node.mli:
```diff
@@ -19,6 +19,12 @@ val external_ : t -> profile:string -> default:Env.t -> Env.t
 val ocaml_flags : t -> profile:string -> expander:Expander.t -> Ocaml_flags.t
 
+val c_flags : t -> profile:string -> expander:Expander.t
+  -> (unit, string list) Build.t
```
Nit: if it's a multiline signature, our style is to put an argument on each line. There's many examples in the codebase of this.

gretay-js replied:
fixed.

rgrinberg commented:
Please sign your commits according to: https://github.com/ocaml/dune/blob/master/CONTRIBUTING.md

gretay-js commented on src/dune_env.ml:
```diff
@@ -39,6 +41,8 @@ module Stanza = struct
     let%map flags = field_oslu "flags"
     and ocamlc_flags = field_oslu "ocamlc_flags"
     and ocamlopt_flags = field_oslu "ocamlopt_flags"
+    and c_flags = field_oslu "c_flags"
```
fixed.

gretay-js commented on src/env_node.ml:
```diff
@@ -109,3 +113,53 @@ let rec ocaml_flags t ~profile ~expander =
+  | None ->
+    let default =
+      match t.inherit_from with
+      | None -> Build.return ([])
```
thanks! fixed.

rgrinberg commented:
I reviewed the latest commits, it looks fine to me. Could you rebase your PR?

> I added a test for (4) as far as I understand it, but the test fails because I do not know how to implement (4). Please can you tell me if the test captures what you meant, and if so how to fix the code.

I'm a bit confused here. As far as I can tell, your test shows the desired behavior is exhibited and the flags from the env are being passed to `(run %{cxx})`. However, I don't see any code that passes these flags to `%{cxx}`. A way to implement this would be to update the bindings for the expander in that directory btw.

Btw, I just realized that we haven't accounted for configurator here. Configurator also compiles C/C++ files so perhaps it should also inherit the environment flags. Currently, it accesses them straight from the context. Do you think it would be desirable to make configurator aware of the setting in env?

Btw: some commits are missing the DCO. Once you've done rebasing, I believe that `$ git rebase master --signoff` should restore the signatures to all commits for you.

rgrinberg commented on test/blackbox-tests/test-cases/env-cflags/run.t:
```diff
@@ -0,0 +1,47 @@
+   (c_flags (-DTEST_C))
+   (cxx_flags (-DTEST_CPP))
+  )
+  $ dune build --profile default run/foo.exe
```
It's a bit simpler to just do `$ dune exec --profile default ./run/foo.exe`. This will build & run the binary and you won't need the dedicated alias anymore.

rgrinberg commented:
Also, let's see what @diml has to say about having configurator respect these c flags:

> Btw, I just realized that we haven't accounted for configurator here. Configurator also compiles C/C++ files so perhaps it should also inherit the environment flags. Currently, it accesses them straight from the context. Do you think it would be desirable to make configurator aware of the setting in env?

diml commented:
I'm not entirely sure. I suppose both choices make sense. We can probably start with the option that's the simplest to implement and see how it goes.

avsm commented:
I think the main usecase for this feature will be to add:

- static analysis cflags like new warnings
- optimisation levels
- include directories to os-specific locations (`/opt/foo/include`) for a particular build

In all of these cases, it seems more useful for configurator to respect the passed in environment flags.  It's just important that these cflags are passed consistently to all the dune-related things that use them.  It would be most useful if configurator could record somewhere in the build tree the _actual_ cflags that it ended up using (i.e. the combination of the env and local cflags)

diml commented:
Ok, that makes sense. However, let's not make it a requirement to merge this PR. It can be done later in a subsequent PR.

gretay-js commented:
I rebased and fixed expected test outputs to check the defaults based on ocamlc -config using some horrendous combination of sed commands. There must be a better way but this works for now and there are other uses of sed in the test suite. 

> I'm a bit confused here. As far as I can tell, your test shows the desired behavior is exhibited and the flags from the env are being passed to `(run %{cxx})`. However, I don't see any code that passes these flags to `%{cxx}`. 

The test fails and I don't know how to fix the code.

> A way to implement this would be to update the bindings for the expander in that directory btw.
I don't Can you please give me a bit more direction for it?

rgrinberg commented:
In pform.ml, `%{cc}` and `%{cxx}` are defined accordingly:

```
      ; "cc"             , strings (context.c_compiler :: cflags)
      ; "cxx"            , strings (context.c_compiler :: cxx_flags)
```

But that's no longer sufficient and we should update the expander for every `Env_node.t`. It's possible to do this by updating the expander with the following set of functions:

* `Expander.add_bindings`
* `Pform.Map.of_list_exn`

Where the flags will be fetched from the current env_node (using the c_flags and cxx_flags functions you've defined) all inside the `Env.expander` function.

rgrinberg commented on test/blackbox-tests/test-cases/env-cflags/run.t:
```diff
@@ -0,0 +1,18 @@
+  $ export STANDARD_C_FLAGS=`ocamlc -config | grep ocamlc_cflags | sed 's/ocamlc_cflags: //'`
+  $ export STANDARD_CXX_FLAGS=`ocamlc -config | grep ocamlc_cflags | sed 's/ocamlc_cflags: //' | sed 's/-std=[^ ]* //' | sed 's/-std=[^$]*//'`
```
I think CI shows that this doesn't work. Dune starts a new sh process for every `$` so the environment variables aren't saved. I'm not sure how to workaround this except for writing your own .sh files that will wrap the command with these substitutions.

rgrinberg commented:
@gretay-js are you still planning on finishing this? Otherwise I can take over. This feature would be really nice to have.

gretay-js commented:
Sorry about the delay. Let me have another go at fixing it.

gretay-js commented:
In commit a98b83023ae24c0730201d8836b0581d4b798bdd, I fixed the test failures that depend on environment variables as you suggested by adding a script that filters output of printenv. 

Regarding bindings for cc and cxx variables that depend on the bindings of c_flags and cxx_flags variables.  Commit 675a0cdae0f122ec437cd3cb22b887ff7b1789c4 attempts to deal with it following your comment earlier, but it doesn't work and doesn't even build. The problem is that Env.cxx_flags functions that I added return (unit, string list) Build.t  and add_bindings expects a string list directly.  
I am sorry I don't know how to fix it without changes all over the code. Can I leave it to you?

rgrinberg commented:
Yeah, it's quite a bit more complicated than I expected. The idea is that we will not be expanding `cc` and `cxx` in a static context anymore. We need to expand them the way we handle macros such as `%{read}`. That is, `%{cc}` and `%{cxx}` should be able to introduce dependencies.

Unfortunately, that means that it will no longer be possible to expand `%{cc}` in static contexts. While that's not so bad, it is technically a breaking change.

However, if we're introducing full support for C/C++ dependencies I wonder if it makes sense to have `%{cc}` and `%{cxx}` at all. Perhaps we should just consider deprecating these variables? @diml would that be acceptable?

diml commented:
We need a replacement first. These variables might be used in contexts that can't be easily changed:

```
(run ./some_program.exe --c-compiler %{cc} ...)
```

So it seems better to me if we keep them. It most places, things that are expanded dynamically can also be expanded statically when they don't require dynamic expansion. Can't we do the same here? This way it wouldn't break anything.

rgrinberg commented:
This PR has been revived in #1860
